GH-37216: [Docs] adding documentation to deal with unreleased allocators

davisusanibar opened the pull request:
### Rationale for this change

To close #37216 

### What changes are included in this PR?

Documentation added to try to catch unreleased allocations.

### Are these changes tested?

Yes

### Are there any user-facing changes?

No
* Closes: #37216

lidavidm reviewed (changes requested):
We don't want to put a _runtime_ dependency on (two versions of) JUnit to all people using Arrow.

davisusanibar commented:
> We don't want to put a _runtime_ dependency on (two versions of) JUnit to all people using Arrow.

I deleted helper classes and added them to the memory management documentation.

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+.. code-block:: java
+
+    //1
+    private static final RootAllocator allocator = new RootAllocator();
```
```suggestion
    private static final BufferAllocator allocator = new RootAllocator();
```

You generally should never declare RootAllocator as the type of a variable or parameter.

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
    |        at RootAllocator.close (RootAllocator.java:29)
    |        at (#8:1)
 
+A further option would be to change the way how allocators are created, we could try the following:
+
+1. Set up a global root allocator that is going to be used in each module,
+2. Create child allocators by using the root allocator,
+3. Give child allocations proper names so that you can detect errors in another class or method,
+4. Ensure that resources are properly closed,
+5. At some strategic point, check if the global allocator is empty,
+6. In case of error, try to review the above-identified class/method allocator bugs.
```
```suggestion
Sometimes, explicitly passing allocators around is difficult. For example, it
can be hard to pass around extra state, like an allocator, through layers of 
existing application or framework code. A global or singleton allocator instance
can be useful here, though it should not be your first choice.

How this works:

1. Set up a global allocator in a singleton class.
2. Provide methods to create child allocators from the global allocator.
3. Give child allocators proper names to make it easier to figure out where
   allocations occurred in case of errors.
4. Ensure that resources are properly closed.
5. Check that the global allocator is empty at some suitable point, such as
   right before program shutdown.
6. If it is not empty, review the above allocation bugs.
```

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+    ...
+    //2
+    public static BufferAllocator getChildAllocator() {
+        return allocator.newChildAllocator(nextChildName(), 0, Long.MAX_VALUEn);
```
```suggestion
        return allocator.newChildAllocator(nextChildName(), 0, Long.MAX_VALUE);
```

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+    ...
+    //3
+    private static String nextChildName() {
+        return "Allocator-Child-" + getClassNameAndMethodName();
```
`getClassNameAndMethodName` isn't defined

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+    }
+    ...
+    //4: Business code
+    try () {
```
```suggestion
    try (BufferAllocator allocator = GlobalAllocator.getChildAllocator()) {
```

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+    ...
+    //5
+    public static void checkGlobalCleanUpResources() {
+        ...
+        !allocator.getChildAllocators().isEmpty();
+        allocator.getAllocatedMemory() != 0;
+        ...
+        throw new IllegalStateException(...);
```
Wrap this in an `if (...) { throw ...; }`

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,49 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+        throw new IllegalStateException(...);
+    }
+    ...
+    //6: Allocator bug detected
+    Review active allocators on: Allocator-Child-com.yourpackage.Test-method
```
Remove this?

lidavidm commented on docs/source/java/memory.rst:
```diff
@@ -288,6 +288,54 @@ Finally, enabling the ``TRACE`` logging level will automatically provide this st
+    }
+    ...
+    //5
+    public static void checkGlobalCleanUpResources() {
+        ...
+        if (!allocator.getChildAllocators().isEmpty()) {
+          throw new IllegalStateException(...)
+          ));
+        } else if (allocator.getAllocatedMemory() != 0) {
+          throw new IllegalStateException(...);
+        }
+    }
```
```suggestion
    public static void checkGlobalCleanUpResources() {
        ...
        if (!allocator.getChildAllocators().isEmpty()) {
          throw new IllegalStateException(...);
        } else if (allocator.getAllocatedMemory() != 0) {
          throw new IllegalStateException(...);
        }
    }
```
